core/backend/src/CodeService.ts

```diff
@@ -0,0 +1,521 @@
+  addAllCodeSpecs: () => Promise<void>;
+
+  /** the BriefcaseDb of this CodeService */
+  readonly briefcase: BriefcaseDb;
+
+  /** the code index for this CodeService */
+  readonly codeIndex: CodeIndex;
+
+  /**
+   * Application-supplied parameters for obtaining the write lock on the container and for reserving new codes.
+   * Applications should set these parameters by adding a listener for `BriefcaseDb.onCodeServiceCreated`
+   * that is called every time a BriefcaseDb that uses code services is opened for write access.
+   */
+  readonly appParams: CodeService.ObtainLockParams & CodeService.AuthorAndOrigin;
+
+  /**
+   * The token that grants access to the cloud container for this CodeService.
+   * It should be established in a listener for `BriefcaseDb.onCodeServiceCreated`, and should be refreshed (via a
+   * timer) before it expires.
+   */
+  sasToken: AccessToken;
+
+  /**
+   * Synchronize the local index with any changes by made by others.
+   * @note This is called automatically whenever any write operation is performed on the code index. It is only necessary to
+   * call this directly if you have not changed the code index recently, but wish to perform a readonly operation and want to
+   * ensure it is up-to-date as of now.
+   * @note There is no guarantee that a readonly index is up-to-date even immediately after calling this method, since others
+   * may be modifying it at any time.
+   */
+  synchronizeWithCloud: () => void;
+
+  /**
+   * Verify that the Code of a to-be-inserted or to-be-updated Element:
+   * 1. has already been reserved,
+   * 2. if the element has a `federationGuid`, it must match the reserved value. If the federationGuid is undefined,
+   * the value from the code index is returned.
+   *
+   * If not, throw an exception. Elements with no CodeValue are ignored.
+   * @note this method is automatically called whenever elements are added or updated by a BriefcaseDb with a CodeService.
+   */
+  verifyCode: (props: CodeService.ElementCodeProps) => void;
+
+  /** Add a new code spec to this code service.
+   * @note This will automatically attempt to obtain, perform the operation, and then release the write lock.
+   */
+  addCodeSpec: (val: CodeService.NameAndJson) => Promise<void>;
+
+  /**
+   * Add all of the codes and code specs from this CodeService's BriefcaseDb into the code index.
+   * @returns the number of codes actually added.
+   * @note It is not necessary to call this method unless the BriefcaseDb somehow becomes out of sync with its CodeService,
+   * for example when migrating iModels to a new code service. It is safe (but relatively expensive) to call this method multiple times, since
+   * any codes or code specs that are already in the index are ignored.
+   * @note This will automatically attempt to obtain, perform the operation, and then release the write lock.
+   */
+  addAllCodes: () => Promise<number>;
+
+  /**
+   * Attempt to reserve a single proposed code.
```

Comment:
Does it throw (a CodeService.Error containing a CodeService.ReserveProblem?) if it cannot reserve the code?